include/rocksdb/utilities/backupable_db.h

```diff
@@ -458,6 +439,71 @@ class BackupEngine {
     options.progress_callback = progress_callback;
     return CreateNewBackup(options, db);
   }
+};
+
+// A backup engine for organizing and managing backups.
+// This class is not user-extensible.
+//
+// This class declaration adds "Write" operations in addition to the
+// operations from BackupEngineAppendOnlyBase and BackupEngineReadOnlyBase.
+//
+// # Concurrency between threads on the same BackupEngine* object
+//
+// As of version 6.20, BackupEngine* operations are generally thread-safe,
+// using a read-write lock, though single-thread operation is still
+// recommended to avoid TOCTOU bugs. Specifically, particular kinds of
+// concurrent operations behave like this:
+//
+// op1\op2| Read  | Append | Write
+// -------|-------|--------|--------
+//   Read | conc  | block  | block
+// Append | block | block  | block
+//  Write | block | block  | block
+//
+// conc = operations safely proceed concurrently
+// block = one of the operations safely blocks until the other completes.
+//   There is generally no guarantee as to which completes first.
+//
+// StopBackup is the only operation that affects an ongoing operation.
+//
+// # Interleaving operations between BackupEngine* objects open on the
+// same backup_dir
+//
+// It is recommended only to have one BackupEngine* object open for a given
+// backup_dir, but it is possible to mix / interleave some operations
+// (regardless of whether they are concurrent) with these caveats:
+//
+// op1\op2|  Open  |  Read  | Append | Write
+// -------|--------|--------|--------|--------
+//   Open | conc   | conc   | atomic | unspec
+//   Read | conc   | conc   | old    | unspec
+// Append | atomic | old    | unspec | unspec
```

Comment:
Note that GC does not delete corrupt backups, so Open is not affected either.

Reply:
> GarbageCollect can be an Append operation according to these rules

Good point. Should we put it (and maybe also `StopBackup()`) in `BackupEngineAppendOnlyBase`?

Reply:
Actually, I think I can argue for putting GarbageCollect in the Append operations. In terms of logical data exposed to the API user, it is a no-op. It merely cleans up obsolete temporary or corrupt data. If some other BackupEngine is appending or writing to the same backup directory, GC could remove temporary data used in that operation, because it appears to be obsolete. So for concurrency between instances, GC is most like an Append operation: does not interfere with reads but does interfere with other appends/writes.